### ISSUE TITLE
Bug: Selector input.-username does not match <input className="field -username" />

### ISSUE THREAD

**Issue by mik01aj** [airbnb/enzyme#547]
This element:

```
<input type="text" maxLength={32} value="" onChange={[Function]} placeholder="Dein Mitgliedsname" className="field -username" />
```

Is not matched by `input.-username` selector.

Btw, `input.field.-username` works, but the former should work too.

Using enzyme `2.3.0`.

**Comment by ljharb** [comment]
Can you link to the spec that shows where class names can start with a hyphen?

**Comment by mik01aj** [comment]
[The CSS 2.1 spec](https://www.w3.org/TR/CSS21/syndata.html#characters) says it shouldn't start with hyphen followed by a digit, but it's not the case here. And class names starting with hyphens are [recommended for component variants in RSCSS](http://rscss.io/variants.html).

**Comment by mik01aj** [comment]
I really needed this so I looked for a workaround and... attribute matching seems broken too. I have a dialog box that contains this: (as reported by `wrapper.debug()`)

```
<Button action={[Function]} icon="facebook" className="-facebook" variant="filled" disabled={false}>
  <button style={[undefined]} className="Button -filled -facebook" disabled={false} onClick={[Function]}>
    <CommonIcon icon="facebook" className="rightIcon" orientation="right">
      <span className="CommonIcon -facebook rightIcon" dangerouslySetInnerHTML={{...}} />
    </CommonIcon>
  </button>
</Button>
```

And I tried to match the btuton with various selectors. `.-facebook` matched the button and the icon inside, but all these:

```
button.-facebook
.Button.-facebook
.Button.-filled.-facebook
button[className="Button -filled -facebook"]
button[class="Button -filled -facebook"]
[className="Button -filled -facebook"]
[class="Button -filled -facebook"]
[className="-facebook"]
[class="-facebook"]
```

Didn't match anything. Now this seems a pretty serious bug. 🐞

**Comment by ljharb** [comment]
Closed in #1086
